sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala

```diff
@@ -0,0 +1,251 @@
+  // bytes read
+  def bytesRead: Long
+}
+
+trait MultiFileReaderFunctions extends Arm {
+  protected def addPartitionValues(
+      batch: Option[ColumnarBatch],
+      inPartitionValues: InternalRow,
+      partitionSchema: StructType): Option[ColumnarBatch] = {
+    if (partitionSchema.nonEmpty) {
+      batch.map { cb =>
+        val partitionValues = inPartitionValues.toSeq(partitionSchema)
+        val partitionScalars = ColumnarPartitionReaderWithPartitionValues
+          .createPartitionValues(partitionValues, partitionSchema)
+        withResource(partitionScalars) { scalars =>
+          ColumnarPartitionReaderWithPartitionValues.addPartitionValues(cb, scalars,
+            GpuColumnVector.extractTypes(partitionSchema))
+        }
+      }
+    } else {
+      batch
+    }
+  }
+
+  protected def fileSystemBytesRead(): Long = {
+    FileSystem.getAllStatistics.asScala.map(_.getThreadStatistics.getBytesRead).sum
+  }
+}
+
+/**
+ * The Abstract base class working as multi-file cloud reading framework
+ * @param conf
+ * @param files PartitionFiles to be read
+ * @param numThreads the number of threads to parallelly read files.
+ * @param maxNumFileProcessed
+ * @param filters push down filters
+ * @param execMetrics
+ */
+abstract class MultiFileCloudPartitionReaderBase(
+    conf: Configuration,
+    files: Array[PartitionedFile],
+    numThreads: Int,
+    maxNumFileProcessed: Int,
+    filters: Array[Filter],
+    execMetrics: Map[String, GpuMetric]) extends PartitionReader[ColumnarBatch] with Logging
+  with ScanWithMetrics with Arm {
+
+  metrics = execMetrics
+
+  protected var maxDeviceMemory: Long = 0
+
+  protected var batch: Option[ColumnarBatch] = None
+  protected var isDone: Boolean = false
+
+  private var filesToRead = 0
+  protected var currentFileHostBuffers: Option[HostMemoryBuffersWithMetaDataBase] = None
+  private var isInitted = false
+  private val tasks = new ConcurrentLinkedQueue[Future[HostMemoryBuffersWithMetaDataBase]]()
+  private val tasksToRun = new Queue[Callable[HostMemoryBuffersWithMetaDataBase]]()
+  private[this] val inputMetrics = TaskContext.get.taskMetrics().inputMetrics
+
+  private def initAndStartReaders(): Unit = {
+    // limit the number we submit at once according to the config if set
+    val limit = math.min(maxNumFileProcessed, files.length)
+    for (i <- 0 until limit) {
+      val file = files(i)
+      // Add these in the order as we got them so that we can make sure
+      // we process them in the same order as CPU would.
+      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(
+        getBatchRunner(file, conf, filters), numThreads))
+    }
+    // queue up any left to add once others finish
+    for (i <- limit until files.length) {
+      val file = files(i)
+      tasksToRun.enqueue(getBatchRunner(file, conf, filters))
+    }
+    isInitted = true
+    filesToRead = files.length
+  }
+
+  /**
+   * file reading logic in a Callable which will be running in a thread pool
```

Comment:
upper case File, finish descriptions

Reply:
> Note we have started burn down for 0.5 so this should eventually be targeted to 0.6 once that branch is available. Looks like the sign off check failed.

Yeah, I will investigate how to abstract coalescing reading which seems more harder than cloud reading.

Reply:
Done